content/advent-2018/split-data-with-cdc.md

```diff
@@ -0,0 +1,232 @@
+
+Let's see how many different chunks we have in our data file byte counting the different hashes:
+```
+$ cat data | go run main.go | cut -f4 | sort | uniq | wc -l
+69
+```
+
+What happens when we feed the program our data twice, how many chunks does it detect? Let's find out:
+
+```
+$ (cat data; cat data) | go run main.go | cut -f4 | sort | uniq | wc -l
+70
+```
+
+Huh, so we only got a single additional chunk. The cause is that for the first round, it'll find the same chunks as before (it's the same data after all) right until the last chunk. In the previous run, the end of the last chunk was determined by the end of the file. Now there's additional data, so the chunk continues. The next cut point will in the second run of the data, it'll be the same as the first cut point, just with some additional data at the beginning of the chunk, so the SHA-256 hash is different. Afterwards, the same chunks follow in the same order.
+
+Changing a few consecutive bytes of data somewhere in most cases also only affects a single chunk, let's write the output to a file, change the data a bit using `sed`, and write another file:
+
+```
+$ cat data | go run main.go > orig.log
+$ cat data | sed 's/EZE8HX/xxxxxx/' | go run main.go > mod.log
+```
+
+The string `EZE8HX` was present in my randomly generated file `data` only once and the `sed` command above changed it to the string `xxxxxx`. When we now compare the two log files using `diff`, we can see that for exactly one chunk the SHA-256 hash has changed, but the other chunks (and the chunk boundaries) stayed the same:
+
+```diff
+$ diff -au orig.log mod.log
+--- orig.log    2018-11-24 13:14:45.265329525 +0100
++++ mod.log     2018-11-24 13:19:32.344681407 +0100
+@@ -55,7 +55,7 @@
+ 83545247       547876  00198b8839f00000        2cbeea803ba79d54[...]
+ 84093123       889631  0007ddacabc00000        a2b6e8651ae1ab69[...]
+ 84982754       2561935 000c53712f500000        bca589959b019a80[...]
+-87544689       2744485 0000a49118b00000        02b125104f06ed85[...]
++87544689       2744485 0000a49118b00000        ce2e43496ae6e7c3[...]
+ 90289174       1167308 00034d6cce700000        ad25a331993d9db7[...]
+ 91456482       1719951 001d2fea8ae00000        ba5153845c5b5228[...]
+ 93176433       1362655 0003009fc1600000        8e7a35e340c10d61[...]
+```
+
+This technique can be used for many other things besides a backup program.
+
+If you're interested in how the `chunker` package is used in restic, there's a [post over at the restic blog](https://restic.net/blog/2015-09-12/restic-foundation1-cdc) which explains this in a bit more detail.
+
+Performance
+===========
+
+Performance in a backup program is important: if it's too slow people tend to skip backup, which is undesirable. The `chunker` package is written in plain Go (so no fancy assembler functions or optimizations). I've optimized the calculations a bit and it's pretty fast already, the package contains some benchmarks:
```

Comment:
> Performance in a backup program is important:

I would rephrase and split it into something like the following:

"Performance is crucial for backup programs. If creating backups is too slow people tend to stop using it.

`chunker` is written in plain Go. It doesn't have any fancy low-level assembler, but the code is already pretty fast, thanks to some optimized calculations. The package has some benchmarks:"

Reply:
LGTM